File: app/models/network.py
from __future__ import annotations
from datetime import datetime, timezone, date
from sqlalchemy import Boolean, Enum, ForeignKey, Integer, String, Text, Numeric, DateTime, UniqueConstraint, Date, BigInteger
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base
from app.models.common import CustomerDeviceStatus, NetDeviceStatus, NetNodeLocationType

class NetNode(Base):
    __tablename__ = "net_nodes"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    division_id: Mapped[int | None] = mapped_column(ForeignKey("divisions.id", ondelete="SET NULL"), nullable=True)
    location_state_id: Mapped[int | None] = mapped_column(ForeignKey("location_states.id", ondelete="SET NULL"), nullable=True)
    location_district_id: Mapped[int | None] = mapped_column(ForeignKey("location_districts.id", ondelete="SET NULL"), nullable=True)
    location_city_id: Mapped[int | None] = mapped_column(ForeignKey("location_cities.id", ondelete="SET NULL"), nullable=True)
    location_street_id: Mapped[int | None] = mapped_column(ForeignKey("location_streets.id", ondelete="SET NULL"), nullable=True)
    street_number: Mapped[str | None] = mapped_column(String(32), nullable=True)
    location_detail: Mapped[str | None] = mapped_column(String(255), nullable=True)
    location_type: Mapped[NetNodeLocationType] = mapped_column(
        Enum(NetNodeLocationType, values_callable=lambda x: [e.value for e in x]),
        default=NetNodeLocationType.other,
        nullable=False,
    )
    latitude: Mapped[float | None] = mapped_column(Numeric(10, 6), nullable=True)
    longitude: Mapped[float | None] = mapped_column(Numeric(10, 6), nullable=True)
    x_1992: Mapped[float | None] = mapped_column(Numeric(12, 2), nullable=True)
    y_1992: Mapped[float | None] = mapped_column(Numeric(12, 2), nullable=True)
    info: Mapped[str | None] = mapped_column(Text, nullable=True)
    node_type: Mapped[str | None] = mapped_column(String(64), nullable=True)
    owner_type: Mapped[str | None] = mapped_column(String(64), nullable=True)
    sidusis_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    has_power: Mapped[bool] = mapped_column(Boolean, default=False)
    has_env_control: Mapped[bool] = mapped_column(Boolean, default=False)
    uke_node_kind: Mapped[str | None] = mapped_column(String(64), nullable=True)
    uke_access_rules: Mapped[str | None] = mapped_column(String(64), nullable=True)

    division: Mapped["Division | None"] = relationship(back_populates="net_nodes")
    location_state: Mapped["LocationState | None"] = relationship()
    location_district: Mapped["LocationDistrict | None"] = relationship()
    location_city: Mapped["LocationCity | None"] = relationship()
    location_street: Mapped["LocationStreet | None"] = relationship()
    
    devices: Mapped[list["NetDevice"]] = relationship(back_populates="net_node")
    outgoing_links: Mapped[list["NetNodeLink"]] = relationship(
        "NetNodeLink", foreign_keys="NetNodeLink.source_node_id", back_populates="source_node"
    )
    incoming_links: Mapped[list["NetNodeLink"]] = relationship(
        "NetNodeLink", foreign_keys="NetNodeLink.target_node_id", back_populates="target_node"
    )

class NetNodeLink(Base):
    __tablename__ = "net_node_links"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    source_node_id: Mapped[int] = mapped_column(ForeignKey("net_nodes.id", ondelete="CASCADE"))
    target_node_id: Mapped[int] = mapped_column(ForeignKey("net_nodes.id", ondelete="CASCADE"))
    medium_type: Mapped[str] = mapped_column(String(32), default="Fiber")
    capacity_mbps: Mapped[int | None] = mapped_column(Integer, nullable=True)
    distance_m: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    info: Mapped[str | None] = mapped_column(String(255), nullable=True)
    source_node: Mapped["NetNode"] = relationship("NetNode", foreign_keys=[source_node_id], back_populates="outgoing_links")
    target_node: Mapped["NetNode"] = relationship("NetNode", foreign_keys=[target_node_id], back_populates="incoming_links")

class NetworkHost(Base):
    __tablename__ = "network_hosts"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), unique=True, nullable=False)
    description: Mapped[str] = mapped_column(Text, default="", nullable=False)
    ip_networks: Mapped[list["IpNetwork"]] = relationship(back_populates="network_host")

class IpNetwork(Base):
    __tablename__ = "ip_networks"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    cidr: Mapped[str] = mapped_column(String(64), nullable=False)
    gateway: Mapped[str | None] = mapped_column(String(64), nullable=True)
    vlan_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    network_host_id: Mapped[int | None] = mapped_column(ForeignKey("network_hosts.id", ondelete="SET NULL"), nullable=True)
    net_device_id: Mapped[int | None] = mapped_column(ForeignKey("net_devices.id", ondelete="SET NULL"), nullable=True)
    net_device_dhcp_server_id: Mapped[int | None] = mapped_column(ForeignKey("net_device_dhcp_servers.id", ondelete="SET NULL"), nullable=True)
    
    network_host: Mapped["NetworkHost | None"] = relationship(back_populates="ip_networks")
    net_devices: Mapped[list["NetDevice"]] = relationship(back_populates="ip_network", foreign_keys="NetDevice.ip_network_id")
    customer_devices: Mapped[list["CustomerDevice"]] = relationship(back_populates="ip_network")
    managing_net_device: Mapped["NetDevice | None"] = relationship(back_populates="managed_ip_networks", foreign_keys=[net_device_id])
    dhcp_server: Mapped["NetDeviceDhcpServer | None"] = relationship(back_populates="ip_networks")


class NetDeviceProducer(Base):
    __tablename__ = "net_device_producers"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), unique=True, nullable=False)
    alternative_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    models: Mapped[list["NetDeviceModel"]] = relationship(back_populates="producer")

class NetDeviceType(Base):
    __tablename__ = "net_device_types"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    models: Mapped[list["NetDeviceModel"]] = relationship(back_populates="device_type")

class NetDeviceModel(Base):
    __tablename__ = "net_device_models"
    __table_args__ = (UniqueConstraint("producer_id", "name", name="uq_net_device_models_producer_name"),)
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    producer_id: Mapped[int] = mapped_column(ForeignKey("net_device_producers.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    alternative_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    type_id: Mapped[int | None] = mapped_column(ForeignKey("net_device_types.id", ondelete="SET NULL"), nullable=True)
    producer: Mapped["NetDeviceProducer"] = relationship(back_populates="models")
    device_type: Mapped["NetDeviceType | None"] = relationship(back_populates="models")
    devices: Mapped[list["NetDevice"]] = relationship(back_populates="net_device_model")

class NetDeviceInterface(Base):
    __tablename__ = "net_device_interfaces"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    net_device_id: Mapped[int] = mapped_column(ForeignKey("net_devices.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    mac_address: Mapped[str | None] = mapped_column(String(32), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    comment: Mapped[str | None] = mapped_column(Text, nullable=True)
    net_device: Mapped["NetDevice"] = relationship(back_populates="interfaces")
    dhcp_servers: Mapped[list["NetDeviceDhcpServer"]] = relationship(back_populates="interface")

class NetDeviceIpPool(Base):
    __tablename__ = "net_device_ip_pools"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    net_device_id: Mapped[int] = mapped_column(ForeignKey("net_devices.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    ranges: Mapped[str] = mapped_column(Text, nullable=False)
    net_device: Mapped["NetDevice"] = relationship(back_populates="ip_pools")

class NetDeviceDhcpServer(Base):
    __tablename__ = "net_device_dhcp_servers"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    net_device_id: Mapped[int] = mapped_column(ForeignKey("net_devices.id", ondelete="CASCADE"), nullable=False)
    interface_id: Mapped[int | None] = mapped_column(ForeignKey("net_device_interfaces.id", ondelete="SET NULL"), nullable=True)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    address_pool_name: Mapped[str | None] = mapped_column(String(128), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    net_device: Mapped["NetDevice"] = relationship(back_populates="dhcp_servers")
    interface: Mapped["NetDeviceInterface | None"] = relationship(back_populates="dhcp_servers")
    ip_networks: Mapped[list["IpNetwork"]] = relationship(back_populates="dhcp_server")

class NetDevice(Base):
    __tablename__ = "net_devices"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    hostname: Mapped[str | None] = mapped_column(String(255), nullable=True)
    serial_number: Mapped[str | None] = mapped_column(String(128), nullable=True)
    mac_address: Mapped[str | None] = mapped_column(String(32), nullable=True)
    management_ip: Mapped[str | None] = mapped_column(String(64), nullable=True)
    device_type: Mapped[str] = mapped_column(String(64), default="other", nullable=False)
    snmp_community: Mapped[str | None] = mapped_column(String(128), nullable=True)
    login_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    ip_network_id: Mapped[int | None] = mapped_column(ForeignKey("ip_networks.id", ondelete="SET NULL"), nullable=True)
    net_node_id: Mapped[int | None] = mapped_column(ForeignKey("net_nodes.id", ondelete="SET NULL"), nullable=True)
    customer_id: Mapped[int | None] = mapped_column(ForeignKey("customers.id", ondelete="SET NULL"), nullable=True)
    net_device_model_id: Mapped[int | None] = mapped_column(ForeignKey("net_device_models.id", ondelete="SET NULL"), nullable=True)
    status: Mapped[NetDeviceStatus] = mapped_column(
        Enum(NetDeviceStatus, values_callable=lambda x: [e.value for e in x]),
        default=NetDeviceStatus.active,
        nullable=False,
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    driver_type: Mapped[str | None] = mapped_column(String(32), nullable=True)
    mgmt_username: Mapped[str | None] = mapped_column(String(64), nullable=True)
    mgmt_password_encrypted: Mapped[str | None] = mapped_column(String(255), nullable=True)
    
    # Monitoring fields
    last_seen: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    is_online: Mapped[bool] = mapped_column(Boolean, default=True)
    latency_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    net_device_model: Mapped["NetDeviceModel | None"] = relationship(back_populates="devices")
    ip_network: Mapped["IpNetwork | None"] = relationship(back_populates="net_devices", foreign_keys=[ip_network_id])
    net_node: Mapped["NetNode | None"] = relationship(back_populates="devices")
    owner_customer: Mapped["Customer | None"] = relationship(back_populates="owned_net_devices", foreign_keys="NetDevice.customer_id")
    customer_devices: Mapped[list["CustomerDevice"]] = relationship(back_populates="net_device", foreign_keys="CustomerDevice.net_device_id")
    interfaces: Mapped[list["NetDeviceInterface"]] = relationship(back_populates="net_device", cascade="all, delete-orphan")
    ip_pools: Mapped[list["NetDeviceIpPool"]] = relationship(back_populates="net_device", cascade="all, delete-orphan")
    dhcp_servers: Mapped[list["NetDeviceDhcpServer"]] = relationship(back_populates="net_device", cascade="all, delete-orphan")
    managed_ip_networks: Mapped[list["IpNetwork"]] = relationship(back_populates="managing_net_device", foreign_keys="IpNetwork.net_device_id")

class CustomerDevice(Base):
    __tablename__ = "customer_devices"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    customer_id: Mapped[int] = mapped_column(ForeignKey("customers.id"), nullable=False)
    name: Mapped[str | None] = mapped_column(String(128), nullable=True)
    hostname: Mapped[str] = mapped_column(String(255), nullable=False)
    login: Mapped[str | None] = mapped_column(String(64), nullable=True)
    passwd: Mapped[str | None] = mapped_column(String(64), nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(64), nullable=True)
    mac_address: Mapped[str | None] = mapped_column(String(32), nullable=True)
    status: Mapped[CustomerDeviceStatus] = mapped_column(Enum(CustomerDeviceStatus), default=CustomerDeviceStatus.active, nullable=False)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    net_device_id: Mapped[int | None] = mapped_column(ForeignKey("net_devices.id", ondelete="SET NULL"), nullable=True)
    ip_network_id: Mapped[int | None] = mapped_column(ForeignKey("ip_networks.id", ondelete="SET NULL"), nullable=True)
    dhcp_server: Mapped[str | None] = mapped_column(String(128), nullable=True)
    dhcp_interface: Mapped[str | None] = mapped_column(String(64), nullable=True)
    customer: Mapped["Customer"] = relationship(back_populates="devices")
    net_device: Mapped["NetDevice | None"] = relationship(back_populates="customer_devices", foreign_keys="CustomerDevice.net_device_id")
    ip_network: Mapped["IpNetwork | None"] = relationship(back_populates="customer_devices")
    groups: Mapped[list["CustomerDeviceGroup"]] = relationship(secondary="node_group_members", back_populates="devices")
    sessions: Mapped[list["CustomerDeviceSession"]] = relationship(back_populates="device", cascade="all, delete-orphan")
    notices: Mapped[list["CustomerDeviceNotice"]] = relationship(back_populates="device", cascade="all, delete-orphan")
    subscriptions: Mapped[list["Subscription"]] = relationship(back_populates="device")

class CustomerDeviceGroup(Base):
    __tablename__ = "node_groups"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(128), unique=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    devices: Mapped[list["CustomerDevice"]] = relationship(secondary="node_group_members", back_populates="groups")

class CustomerDeviceSession(Base):
    __tablename__ = "node_sessions"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    device_id: Mapped[int] = mapped_column(ForeignKey("customer_devices.id", ondelete="CASCADE"), nullable=False)
    started_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    ended_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(64), nullable=True)
    source: Mapped[str] = mapped_column(String(32), default="manual", nullable=False)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    device: Mapped["CustomerDevice"] = relationship(back_populates="sessions")

class CustomerDeviceNotice(Base):
    __tablename__ = "node_notices"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    device_id: Mapped[int] = mapped_column(ForeignKey("customer_devices.id", ondelete="CASCADE"), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    body: Mapped[str | None] = mapped_column(Text, nullable=True)
    valid_until: Mapped[date | None] = mapped_column(Date, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    device: Mapped["CustomerDevice"] = relationship(back_populates="notices")

class TrafficStat(Base):
    __tablename__ = "traffic_stats"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    device_id: Mapped[int | None] = mapped_column(ForeignKey("customer_devices.id", ondelete="SET NULL"), nullable=True)
    period_start: Mapped[date] = mapped_column(Date, nullable=False)
    period_end: Mapped[date] = mapped_column(Date, nullable=False)
    bytes_in: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    bytes_out: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)

File: crm-portal-ts/src/models/network.ts
import {
    Entity,
    PrimaryGeneratedColumn,
    Column,
    ManyToOne,
    JoinColumn,
    OneToMany,
    OneToOne,
} from "typeorm";
import { Customer } from "./customer.js";
import { CustomerDeviceStatus, NetDeviceStatus, NetNodeLocationType } from "./common.js";

@Entity("net_nodes")
export class NetNode {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "varchar", length: 255 })
    name!: string;

    @Column({ type: "varchar", name: "location_detail", length: 255, nullable: true })
    locationDetail?: string;

    @Column({
        type: "simple-enum",
        name: "location_type",
        enum: NetNodeLocationType,
        default: NetNodeLocationType.other,
    })
    locationType!: NetNodeLocationType;

    @Column({ type: "varchar", name: "node_type", length: 64, nullable: true })
    nodeType?: string;

    @Column({ type: "varchar", name: "owner_type", length: 64, nullable: true })
    ownerType?: string;

    @Column({ type: "real", nullable: true })
    latitude?: number;

    @Column({ type: "real", nullable: true })
    longitude?: number;

    @Column({ type: "real", name: "x_1992", nullable: true })
    x1992?: number;

    @Column({ type: "real", name: "y_1992", nullable: true })
    y1992?: number;

    @Column({ type: "boolean", name: "has_power", default: false })
    hasPower!: boolean;

    @Column({ type: "boolean", name: "has_env_control", default: false })
    hasEnvControl!: boolean;

    @Column({ type: "text", nullable: true })
    info?: string;

    @OneToMany(() => NetDevice, (device) => device.netNode)
    devices!: NetDevice[];
}

@Entity("ip_networks")
export class IpNetwork {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "varchar", length: 128 })
    name!: string;

    @Column({ type: "varchar", length: 64 })
    cidr!: string;

    @Column({ type: "varchar", length: 64, nullable: true })
    gateway?: string;

    @Column({ type: "integer", name: "vlan_id", nullable: true })
    vlanId?: number;

    @Column({ type: "text", nullable: true })
    description?: string;

    @Column({ type: "boolean", default: true })
    active!: boolean;

    @OneToMany(() => NetDevice, (device) => device.ipNetwork)
    netDevices!: NetDevice[];

    @OneToMany(() => CustomerDevice, (device) => device.ipNetwork)
    customerDevices!: CustomerDevice[];
}

@Entity("net_devices")
export class NetDevice {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "varchar", length: 128 })
    name!: string;

    @Column({ type: "varchar", length: 255, nullable: true })
    hostname?: string;

    @Column({ type: "varchar", name: "serial_number", length: 128, nullable: true })
    serialNumber?: string;

    @Column({ type: "varchar", name: "mac_address", length: 32, nullable: true })
    macAddress?: string;

    @Column({ type: "varchar", name: "management_ip", length: 64, nullable: true })
    managementIp?: string;

    @Column({ type: "varchar", name: "device_type", length: 64, default: "other" })
    deviceType!: string;

    @Column({
        type: "simple-enum",
        enum: NetDeviceStatus,
        default: NetDeviceStatus.active,
    })
    status!: NetDeviceStatus;

    @Column({ type: "integer", name: "ip_network_id", nullable: true })
    ipNetworkId?: number;

    @Column({ type: "integer", name: "net_node_id", nullable: true })
    netNodeId?: number;

    @Column({ type: "integer", name: "customer_id", nullable: true })
    customerId?: number;

    @Column({ type: "text", nullable: true })
    notes?: string;

    @ManyToOne(() => IpNetwork, (network) => network.netDevices, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "ip_network_id" })
    ipNetwork?: IpNetwork | null;

    @ManyToOne(() => NetNode, (node) => node.devices, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "net_node_id" })
    netNode?: NetNode | null;

    @ManyToOne(() => Customer, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "customer_id" })
    customer?: Customer | null;

    @OneToOne(() => NetworkDeviceAccessProfile, (profile) => profile.netDevice)
    accessProfile?: NetworkDeviceAccessProfile | null;

    @OneToMany(() => NetworkDiscoverySession, (session) => session.netDevice)
    discoverySessions!: NetworkDiscoverySession[];
}

@Entity("customer_devices")
export class CustomerDevice {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "customer_id" })
    customerId!: number;

    @Column({ type: "varchar", length: 128, nullable: true })
    name?: string;

    @Column({ type: "varchar", length: 255 })
    hostname!: string;

    @Column({ type: "varchar", length: 64, nullable: true })
    login?: string;

    @Column({ type: "varchar", length: 64, nullable: true })
    passwd?: string;

    @Column({ type: "varchar", name: "ip_address", length: 64, nullable: true })
    ipAddress?: string;

    @Column({ type: "varchar", name: "mac_address", length: 32, nullable: true })
    macAddress?: string;

    @Column({
        type: "simple-enum",
        enum: CustomerDeviceStatus,
        default: CustomerDeviceStatus.active
    })
    status!: CustomerDeviceStatus;

    @Column({ type: "text", nullable: true })
    notes?: string;

    @Column({ type: "integer", name: "net_device_id", nullable: true })
    netDeviceId?: number;

    @Column({ type: "integer", name: "ip_network_id", nullable: true })
    ipNetworkId?: number;

    @Column({ type: "varchar", name: "remote_vendor", length: 64, nullable: true })
    remoteVendor?: string;

    @Column({ type: "varchar", name: "remote_serial_number", length: 128, nullable: true })
    remoteSerialNumber?: string;

    @Column({ type: "integer", name: "remote_olt", nullable: true })
    remoteOlt?: number;

    @Column({ type: "integer", name: "remote_onu", nullable: true })
    remoteOnu?: number;

    @Column({ type: "varchar", name: "remote_port", length: 64, nullable: true })
    remotePort?: string;

    @Column({ type: "varchar", name: "remote_profile_name", length: 128, nullable: true })
    remoteProfileName?: string;

    @Column({ type: "real", name: "remote_rx_power_dbm", nullable: true })
    remoteRxPowerDbm?: number;

    @Column({ type: "varchar", name: "installation_state", length: 128, nullable: true })
    installationState?: string;

    @Column({ type: "varchar", name: "installation_county", length: 128, nullable: true })
    installationCounty?: string;

    @Column({ type: "varchar", name: "installation_city", length: 128, nullable: true })
    installationCity?: string;

    @Column({ type: "varchar", name: "installation_street", length: 255, nullable: true })
    installationStreet?: string;

    @Column({ type: "varchar", name: "installation_street_number", length: 32, nullable: true })
    installationStreetNumber?: string;

    @Column({ type: "varchar", name: "installation_apartment_number", length: 32, nullable: true })
    installationApartmentNumber?: string;

    @Column({ type: "varchar", name: "installation_postal_code", length: 32, nullable: true })
    installationPostalCode?: string;

    @Column({ type: "varchar", name: "installation_country", length: 128, nullable: true })
    installationCountry?: string;

    @Column({ type: "varchar", name: "location_description", length: 255, nullable: true })
    locationDescription?: string;

    @ManyToOne(() => Customer, (customer) => customer.devices)
    @JoinColumn({ name: "customer_id" })
    customer!: Customer;

    @ManyToOne(() => NetDevice, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "net_device_id" })
    netDevice?: NetDevice | null;

    @ManyToOne(() => IpNetwork, (network) => network.customerDevices, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "ip_network_id" })
    ipNetwork?: IpNetwork | null;
}

@Entity("network_device_access_profiles")
export class NetworkDeviceAccessProfile {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "net_device_id", unique: true })
    netDeviceId!: number;

    @Column({ type: "varchar", length: 64 })
    driver!: string;

    @Column({ type: "varchar", length: 255 })
    host!: string;

    @Column({ type: "integer", nullable: true })
    port?: number;

    @Column({ type: "varchar", length: 128 })
    username!: string;

    @Column({ type: "text", name: "password_ciphertext" })
    passwordCiphertext!: string;

    @Column({ type: "text", name: "enable_password_ciphertext", nullable: true })
    enablePasswordCiphertext?: string;

    @Column({ type: "text", name: "metadata_json", nullable: true })
    metadataJson?: string;

    @Column({ type: "datetime", name: "created_at", default: () => "CURRENT_TIMESTAMP" })
    createdAt!: Date;

    @Column({ type: "datetime", name: "updated_at", default: () => "CURRENT_TIMESTAMP" })
    updatedAt!: Date;

    @ManyToOne(() => NetDevice, (netDevice) => netDevice.accessProfile, { onDelete: "CASCADE" })
    @JoinColumn({ name: "net_device_id" })
    netDevice!: NetDevice;
}

@Entity("network_discovery_sessions")
export class NetworkDiscoverySession {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "net_device_id" })
    netDeviceId!: number;

    @Column({ type: "integer", name: "access_profile_id", nullable: true })
    accessProfileId?: number;

    @Column({ type: "varchar", length: 64 })
    driver!: string;

    @Column({ type: "varchar", length: 32, default: "running" })
    status!: string;

    @Column({ type: "text", nullable: true })
    summary?: string;

    @Column({ type: "text", name: "error_message", nullable: true })
    errorMessage?: string;

    @Column({ type: "datetime", name: "started_at", default: () => "CURRENT_TIMESTAMP" })
    startedAt!: Date;

    @Column({ type: "datetime", name: "finished_at", nullable: true })
    finishedAt?: Date;

    @ManyToOne(() => NetDevice, (netDevice) => netDevice.discoverySessions, { onDelete: "CASCADE" })
    @JoinColumn({ name: "net_device_id" })
    netDevice!: NetDevice;

    @ManyToOne(() => NetworkDeviceAccessProfile, { nullable: true, onDelete: "SET NULL" })
    @JoinColumn({ name: "access_profile_id" })
    accessProfile?: NetworkDeviceAccessProfile | null;

    @OneToMany(() => NetworkDiscoveryRecord, (record) => record.session)
    records!: NetworkDiscoveryRecord[];
}

@Entity("network_discovery_records")
export class NetworkDiscoveryRecord {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "session_id" })
    sessionId!: number;

    @Column({ type: "integer", name: "net_device_id" })
    netDeviceId!: number;

    @Column({ type: "varchar", name: "record_kind", length: 64 })
    recordKind!: string;

    @Column({ type: "varchar", name: "external_key", length: 255, nullable: true })
    externalKey?: string;

    @Column({ type: "varchar", length: 255, nullable: true })
    hostname?: string;

    @Column({ type: "varchar", name: "ip_address", length: 64, nullable: true })
    ipAddress?: string;

    @Column({ type: "varchar", name: "mac_address", length: 32, nullable: true })
    macAddress?: string;

    @Column({ type: "varchar", name: "serial_number", length: 128, nullable: true })
    serialNumber?: string;

    @Column({ type: "varchar", name: "record_status", length: 64, nullable: true })
    recordStatus?: string;

    @Column({ type: "varchar", name: "profile_name", length: 128, nullable: true })
    profileName?: string;

    @Column({ type: "text", name: "fail_reason", nullable: true })
    failReason?: string;

    @Column({ type: "varchar", name: "remote_port", length: 64, nullable: true })
    remotePort?: string;

    @Column({ type: "integer", name: "remote_vlan_id", nullable: true })
    remoteVlanId?: number;

    @Column({ type: "integer", name: "remote_olt", nullable: true })
    remoteOlt?: number;

    @Column({ type: "integer", name: "remote_onu", nullable: true })
    remoteOnu?: number;

    @Column({ type: "integer", name: "distance_meters", nullable: true })
    distanceMeters?: number;

    @Column({ type: "real", name: "rx_power_dbm", nullable: true })
    rxPowerDbm?: number;

    @Column({ type: "text", name: "payload_json", nullable: true })
    payloadJson?: string;

    @Column({ type: "datetime", name: "created_at", default: () => "CURRENT_TIMESTAMP" })
    createdAt!: Date;

    @ManyToOne(() => NetworkDiscoverySession, (session) => session.records, { onDelete: "CASCADE" })
    @JoinColumn({ name: "session_id" })
    session!: NetworkDiscoverySession;

    @ManyToOne(() => NetDevice, { onDelete: "CASCADE" })
    @JoinColumn({ name: "net_device_id" })
    netDevice!: NetDevice;
}
